### DISCUSSION TITLE
Feature request - MQTT Client Certificate Authentication (MTLs)

### DISCUSSION THREAD

**Discussion by Daniel-dev22** [cesanta/mongoose#2865] [Feature requests]
I took a look at the mqtt configuration in this library and I don't see any mechanism to connect to mqtt via certificate authentication otherwise known as MTLs. Unless I'm mistaken.

This addition would greatly enhance security. It should be implemented in a way that it's optional to use but for those that have certificate authentication setup this provides additional security.

Quick background on client certificate authentication.
https://www.ibm.com/docs/en/ibm-mq/9.2?topic=authentication-mqtt-client-using-tls

I'm more familiar with the python paho mqtt implementation so I'll provide some links to those projects where I helped get MTLs support added as a reference.

jgyates/genmon#1006
bkbilly/lnxlink#87

Happy to help answer any questions on how this works if needed as well.

**Comment by scaprile** [comment]
We do support client certificate authentication by using two-way TLS. Authentication is done at the TLS level, either by our built-in TLS or by OpenSSL, WolfSSL, MbedTLS.

**Reply by Daniel-dev22** [reply]
@scaprile I'm referring to MQTT Certificate Authentication though.

Are you saying you support that?

**Reply by Daniel-dev22** [reply]
This is paho mqtt implementation of client certiticate authentication 
https://eclipse.dev/paho/files/paho.mqtt.python/html/client.html#paho.mqtt.client.Client.tls_set

**Reply by scaprile** [reply]
I think you are confusing what a computer application does with how a communications stack works.
AFAIK there is no such thing as MQTT Certificate Authentication, the security chapter in both MQTT 3.1.1 and MQTT 5.0 is non-normative, and they suggest using TLS. What you are referring to is TLS mutual authentication, also known as 2-way authentication.
MQTT runs on top of TCP. If you want, you can run an intermediate session layer, TLS, and configure to authenticate the server, or both server and client.
And, BTW, Mongoose is not a computer application.
https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html
https://docs.oasis-open.org/mqtt/mqtt/v5.0/mqtt-v5.0.html
https://mongoose.ws/documentation/tutorials/tls/
https://mongoose.ws/documentation/tutorials/tls/#two-way-tls
https://github.com/cesanta/mongoose/blob/b566b8b45a2a54224c48ba685b3aacd3267219fc/tutorials/mqtt/mqtt-client/main.c#L31-L36
just add the client cert as described in the tutorial above:
```
  struct mg_tls_opts opts = {.ca = mg_unpacked("/certs/ca.pem"),
                             .cert = mg_unpacked("/certs/client_cert.pem"),
                             .key = mg_unpacked("/certs/client_key.pem")};
```

**Reply by Daniel-dev22** [reply]
MQTT itself is handling certificate authentication to validate clients auth. Not an external application like a reverse proxy or otherwise.

I'm using client certificate authentication with mqtt right now.

This is what my broker configuration looks like

```
listener 8883 0.0.0.0

tls_version tlsv1.3

cafile /mosquitto/cert/ca.pem
keyfile /mosquitto/cert/client-key.pem
certfile /mosquitto/cert/client.pem
require_certificate true
```
When the client connects to the broker it must have a CA cert, client cert and client key. Then the mqtt broker validates the client cert + client key and the client validates the certificate presented by the broker with the CA cert.

This is what I'm referring to.

https://www.hivemq.com/blog/mqtt-security-fundamentals-x509-client-certificate-authentication/

http://www.steves-internet-guide.com/creating-and-using-client-certificates-with-mqtt-and-mosquitto/

**Reply by scaprile** [reply]
No, that is MQTT over TLS over TCP. and we already support that. You confuse what a computer program does with how systems and communications stacks work.
It is not MQTT that does that, it is the application that starts a TLS session when accepting a TCP connection at port 8883
The client responds and starts the TLS session and after that a regular MQTT login takes place.

See how we handle AWS https://mongoose.ws/documentation/tutorials/mqtt/mqtt-client-aws-iot/ (in addition to all linked and explained above)

Furthermore, this is a one-way example, it can be done mutual/two-way by following indications above or observing the AWS tutorial. We don't connect to HiveMQ because it doesn't yet support TLS1.3, while our built-in stack is 1.3. You can also use MbedTLS or OpenSSL or WolfSSL, as I already told you.
https://github.com/cesanta/mongoose/blob/b566b8b45a2a54224c48ba685b3aacd3267219fc/tutorials/tcpip/pcap-driver/main.c#L9-L86
https://github.com/cesanta/mongoose/blob/b566b8b45a2a54224c48ba685b3aacd3267219fc/tutorials/tcpip/pcap-driver/main.c#L172-L175
